File: src/tqi/raptor/timetable.py
"""Build RAPTOR timetable data structures from parsed GTFS.

Two representations are maintained:
- RaptorTimetable: Python-object version (used for building, testing, pickling)
- FlatTimetable: Numba-compatible flat numpy arrays (used by the JIT engine)
"""

from dataclasses import dataclass
from math import radians, cos

import numpy as np
from scipy.spatial import cKDTree

from tqi.config import EARTH_RADIUS_M, MAX_TRANSFER_WALK_M, WALK_SPEED_M_PER_MIN
from tqi.gtfs.parse import GTFSFeed


@dataclass
class RaptorTimetable:
    """Pre-processed timetable for RAPTOR routing."""
    n_stops: int
    stop_ids: list[str]
    stop_id_to_idx: dict[str, int]

    n_patterns: int
    pattern_stops: list[list[int]]
    pattern_trips: list[np.ndarray]  # pattern_idx → (n_trips, n_stops_in_pattern, 2)

    stop_to_patterns: list[list[tuple[int, int]]]
    transfers: list[list[tuple[int, float]]]


@dataclass
class FlatTimetable:
    """Numba-compatible flat arrays for the JIT RAPTOR engine.

    All jagged lists are flattened into contiguous arrays with offset tables.
    """
    n_stops: int
    n_patterns: int

    # Pattern stops: flat array of stop indices + offsets
    ps_data: np.ndarray     # int32 — concatenated stop indices for all patterns
    ps_offsets: np.ndarray   # int32 — shape (n_patterns+1,), ps_data[ps_offsets[p]:ps_offsets[p+1]]

    # Trip times: flat array of (arr, dep) pairs + offsets per pattern
    # For pattern p with S stops and T trips: tt_data[offset : offset + T*S*2]
    # reshaped to (T, S, 2)
    tt_data: np.ndarray     # int32 — all trip arrival/departure times
    tt_offsets: np.ndarray   # int32 — shape (n_patterns+1,)
    tt_n_trips: np.ndarray   # int32 — shape (n_patterns,), number of trips per pattern
    tt_n_stops: np.ndarray   # int32 — shape (n_patterns,), number of stops per pattern

    # Stop → patterns: flat (pattern_idx, position) pairs + offsets
    sp_data: np.ndarray     # int32 — shape (N, 2), columns: [pattern_idx, position]
    sp_offsets: np.ndarray   # int32 — shape (n_stops+1,)

    # Transfers: flat (target_stop, walk_min_x100) pairs + offsets
    tr_data: np.ndarray     # int32 — shape (N, 2), columns: [target_stop_idx, walk_min*100]
    tr_offsets: np.ndarray   # int32 — shape (n_stops+1,)


def flatten_timetable(tt: RaptorTimetable) -> FlatTimetable:
    """Convert a RaptorTimetable to Numba-compatible flat arrays."""
    # Pattern stops
    ps_parts = []
    ps_offsets = [0]
    for stops in tt.pattern_stops:
        ps_parts.extend(stops)
        ps_offsets.append(len(ps_parts))
    ps_data = np.array(ps_parts, dtype=np.int32)
    ps_offsets = np.array(ps_offsets, dtype=np.int32)

    # Trip times
    tt_parts = []
    tt_offsets = [0]
    tt_n_trips = []
    tt_n_stops = []
    for trips_arr in tt.pattern_trips:
        n_t, n_s, _ = trips_arr.shape
        tt_n_trips.append(n_t)
        tt_n_stops.append(n_s)
        tt_parts.append(trips_arr.reshape(-1))
        tt_offsets.append(tt_offsets[-1] + n_t * n_s * 2)
    tt_data = np.concatenate(tt_parts).astype(np.int32) if tt_parts else np.empty(0, dtype=np.int32)
    tt_offsets = np.array(tt_offsets, dtype=np.int32)
    tt_n_trips_arr = np.array(tt_n_trips, dtype=np.int32)
    tt_n_stops_arr = np.array(tt_n_stops, dtype=np.int32)

    # Stop → patterns
    sp_parts = []
    sp_offsets = [0]
    for entries in tt.stop_to_patterns:
        for pidx, pos in entries:
            sp_parts.append((pidx, pos))
        sp_offsets.append(len(sp_parts))
    sp_data = np.array(sp_parts, dtype=np.int32).reshape(-1, 2) if sp_parts else np.empty((0, 2), dtype=np.int32)
    sp_offsets = np.array(sp_offsets, dtype=np.int32)

    # Transfers
    tr_parts = []
    tr_offsets = [0]
    for entries in tt.transfers:
        for target, walk_min in entries:
            # Store walk_min as int (×100 for precision)
            tr_parts.append((target, int(walk_min * 100)))
        tr_offsets.append(len(tr_parts))
    tr_data = np.array(tr_parts, dtype=np.int32).reshape(-1, 2) if tr_parts else np.empty((0, 2), dtype=np.int32)
    tr_offsets = np.array(tr_offsets, dtype=np.int32)

    return FlatTimetable(
        n_stops=tt.n_stops,
        n_patterns=tt.n_patterns,
        ps_data=ps_data,
        ps_offsets=ps_offsets,
        tt_data=tt_data,
        tt_offsets=tt_offsets,
        tt_n_trips=tt_n_trips_arr,
        tt_n_stops=tt_n_stops_arr,
        sp_data=sp_data,
        sp_offsets=sp_offsets,
        tr_data=tr_data,
        tr_offsets=tr_offsets,
    )


def build_raptor_timetable(feed: GTFSFeed) -> RaptorTimetable:
    """Convert a filtered GTFSFeed into RAPTOR data structures."""
    stop_ids = sorted(feed.stops["stop_id"].unique())
    stop_id_to_idx = {sid: i for i, sid in enumerate(stop_ids)}
    n_stops = len(stop_ids)

    st = feed.stop_times.sort_values(["trip_id", "stop_sequence"])

    trip_sequences: dict[str, list[tuple[int, int, int]]] = {}
    for _, row in st.iterrows():
        sid = row["stop_id"]
        if sid not in stop_id_to_idx:
            continue
        tid = row["trip_id"]
        if tid not in trip_sequences:
            trip_sequences[tid] = []
        trip_sequences[tid].append((
            stop_id_to_idx[sid],
            int(row["arrival_min"]),
            int(row["departure_min"]),
        ))

    pattern_key_to_idx: dict[tuple[int, ...], int] = {}
    pattern_stops: list[list[int]] = []
    pattern_trip_data: list[list[list[tuple[int, int]]]] = []

    for tid, seq in trip_sequences.items():
        stop_key = tuple(s[0] for s in seq)
        if stop_key not in pattern_key_to_idx:
            pattern_key_to_idx[stop_key] = len(pattern_stops)
            pattern_stops.append(list(stop_key))
            pattern_trip_data.append([])
        pidx = pattern_key_to_idx[stop_key]
        pattern_trip_data[pidx].append([(s[1], s[2]) for s in seq])

    pattern_trips: list[np.ndarray] = []
    for trips_for_pattern in pattern_trip_data:
        arr = np.array(trips_for_pattern, dtype=np.int32)
        order = np.argsort(arr[:, 0, 1])
        pattern_trips.append(arr[order])

    n_patterns = len(pattern_stops)

    stop_to_patterns: list[list[tuple[int, int]]] = [[] for _ in range(n_stops)]
    for pidx, stops in enumerate(pattern_stops):
        for pos, sidx in enumerate(stops):
            stop_to_patterns[sidx].append((pidx, pos))

    # Build transfer edges
    stop_lats = np.array([
        feed.stops.loc[feed.stops["stop_id"] == sid, "stop_lat"].iloc[0]
        for sid in stop_ids
    ], dtype=np.float64)
    stop_lons = np.array([
        feed.stops.loc[feed.stops["stop_id"] == sid, "stop_lon"].iloc[0]
        for sid in stop_ids
    ], dtype=np.float64)

    center_lat_rad = radians(stop_lats.mean())
    stop_xy = np.column_stack([
        np.radians(stop_lons) * EARTH_RADIUS_M * cos(center_lat_rad),
        np.radians(stop_lats) * EARTH_RADIUS_M,
    ])
    tree = cKDTree(stop_xy)

    stop_pattern_set: list[set[int]] = [set() for _ in range(n_stops)]
    for pidx, stops in enumerate(pattern_stops):
        for sidx in stops:
            stop_pattern_set[sidx].add(pidx)

    transfers: list[list[tuple[int, float]]] = [[] for _ in range(n_stops)]
    for sidx in range(n_stops):
        nearby = tree.query_ball_point(stop_xy[sidx], r=MAX_TRANSFER_WALK_M)
        for nidx in nearby:
            if nidx == sidx:
                continue
            if stop_pattern_set[nidx] == stop_pattern_set[sidx]:
                continue
            dx = stop_xy[nidx, 0] - stop_xy[sidx, 0]
            dy = stop_xy[nidx, 1] - stop_xy[sidx, 1]
            dist_m = (dx**2 + dy**2) ** 0.5
            walk_min = dist_m / WALK_SPEED_M_PER_MIN
            transfers[sidx].append((nidx, walk_min))

    print(f"RAPTOR timetable: {n_stops} stops, {n_patterns} patterns, "
          f"{sum(len(t) for t in pattern_trips)} trips")

    return RaptorTimetable(
        n_stops=n_stops,
        stop_ids=stop_ids,
        stop_id_to_idx=stop_id_to_idx,
        n_patterns=n_patterns,
        pattern_stops=pattern_stops,
        pattern_trips=pattern_trips,
        stop_to_patterns=stop_to_patterns,
        transfers=transfers,
    )

File: go/internal/raptor/timetable.go
package raptor

import (
	"math"
	"sort"
	"strings"

	"github.com/jasondyck/chwk-tqi/internal/config"
	"github.com/jasondyck/chwk-tqi/internal/geo"
	"github.com/jasondyck/chwk-tqi/internal/gtfs"
)

type stopEntry struct {
	stopIdx int
	arrMin  int
	depMin  int
}

// BuildTimetable converts a GTFS feed into a RAPTOR timetable.
func BuildTimetable(feed *gtfs.Feed) *Timetable {
	// 1. Index stops: collect unique stop IDs from stop_times, sort, create map.
	stopSet := make(map[string]struct{})
	for i := range feed.StopTimes {
		stopSet[feed.StopTimes[i].StopID] = struct{}{}
	}
	stopIDs := make([]string, 0, len(stopSet))
	for id := range stopSet {
		stopIDs = append(stopIDs, id)
	}
	sort.Strings(stopIDs)

	stopIDToIdx := make(map[string]int, len(stopIDs))
	for i, id := range stopIDs {
		stopIDToIdx[id] = i
	}
	nStops := len(stopIDs)

	// 2. Sort stop_times by trip_id then stop_sequence.
	stCopy := make([]gtfs.StopTime, len(feed.StopTimes))
	copy(stCopy, feed.StopTimes)
	sort.Slice(stCopy, func(i, j int) bool {
		if stCopy[i].TripID != stCopy[j].TripID {
			return stCopy[i].TripID < stCopy[j].TripID
		}
		return stCopy[i].StopSequence < stCopy[j].StopSequence
	})

	// 3. Group by trip: trip -> [(stopIdx, arrMin, depMin)]
	tripStops := make(map[string][]stopEntry)
	for _, st := range stCopy {
		idx, ok := stopIDToIdx[st.StopID]
		if !ok {
			continue
		}
		tripStops[st.TripID] = append(tripStops[st.TripID], stopEntry{
			stopIdx: idx,
			arrMin:  st.ArrivalMin,
			depMin:  st.DepartureMin,
		})
	}

	// 4. Group trips into patterns (same sequence of stop indices).
	//    Pattern key = comma-separated stop indices.
	type tripData struct {
		tripID string
		stops  []stopEntry
	}

	patternMap := make(map[string][]tripData)
	for tripID, stops := range tripStops {
		key := patternKey(stops)
		patternMap[key] = append(patternMap[key], tripData{tripID: tripID, stops: stops})
	}

	// Sort pattern keys for deterministic output.
	patternKeys := make([]string, 0, len(patternMap))
	for k := range patternMap {
		patternKeys = append(patternKeys, k)
	}
	sort.Strings(patternKeys)

	nPatterns := len(patternKeys)
	patternStops := make([][]int, nPatterns)
	patternTrips := make([]PatternTripData, nPatterns)

	for pi, key := range patternKeys {
		trips := patternMap[key]
		nStopsInPattern := len(trips[0].stops)

		// Build stop index list for this pattern.
		ps := make([]int, nStopsInPattern)
		for j, se := range trips[0].stops {
			ps[j] = se.stopIdx
		}
		patternStops[pi] = ps

		// 5. Sort trips within each pattern by departure at first stop.
		sort.Slice(trips, func(a, b int) bool {
			return trips[a].stops[0].depMin < trips[b].stops[0].depMin
		})

		// Build flat trip data.
		ptd := PatternTripData{
			NTrips: len(trips),
			NStops: nStopsInPattern,
			Data:   make([]int, len(trips)*nStopsInPattern*2),
		}
		for ti, td := range trips {
			for si, se := range td.stops {
				ptd.Data[(ti*nStopsInPattern+si)*2] = se.arrMin
				ptd.Data[(ti*nStopsInPattern+si)*2+1] = se.depMin
			}
		}
		patternTrips[pi] = ptd
	}

	// 6. Build stopToPatterns.
	stopToPatterns := make([][]StopPatternEntry, nStops)
	for pi, ps := range patternStops {
		for pos, stopIdx := range ps {
			stopToPatterns[stopIdx] = append(stopToPatterns[stopIdx], StopPatternEntry{
				PatternIdx: pi,
				Position:   pos,
			})
		}
	}

	// 7. Build transfers.
	transfers := buildTransfers(feed, stopIDs, stopIDToIdx, nStops)

	return &Timetable{
		NStops:         nStops,
		StopIDs:        stopIDs,
		StopIDToIdx:    stopIDToIdx,
		NPatterns:      nPatterns,
		PatternStops:   patternStops,
		PatternTrips:   patternTrips,
		StopToPatterns: stopToPatterns,
		Transfers:      transfers,
	}
}

// patternKey creates a string key from a sequence of stop entries.
func patternKey(stops []stopEntry) string {
	parts := make([]string, len(stops))
	for i, s := range stops {
		parts[i] = strings.Repeat("0", 6) // placeholder
		// Use a simple int-to-string for the key.
		parts[i] = intToStr(s.stopIdx)
	}
	return strings.Join(parts, ",")
}

func intToStr(n int) string {
	if n == 0 {
		return "0"
	}
	s := ""
	neg := n < 0
	if neg {
		n = -n
	}
	for n > 0 {
		s = string(rune('0'+n%10)) + s
		n /= 10
	}
	if neg {
		s = "-" + s
	}
	return s
}

// buildTransfers creates walking transfers between nearby stops.
func buildTransfers(feed *gtfs.Feed, stopIDs []string, stopIDToIdx map[string]int, nStops int) [][]Transfer {
	transfers := make([][]Transfer, nStops)

	if nStops == 0 {
		return transfers
	}

	// Build lat/lon arrays indexed by our stop index.
	stopByID := make(map[string]*gtfs.Stop, len(feed.Stops))
	for i := range feed.Stops {
		stopByID[feed.Stops[i].StopID] = &feed.Stops[i]
	}

	lats := make([]float64, nStops)
	lons := make([]float64, nStops)
	for i, id := range stopIDs {
		s, ok := stopByID[id]
		if ok {
			lats[i] = s.StopLat
			lons[i] = s.StopLon
		}
	}

	// Project to XY.
	meanLat := 0.0
	for _, lat := range lats {
		meanLat += lat
	}
	meanLat /= float64(nStops)
	centerLatRad := meanLat * math.Pi / 180.0

	xs, ys := geo.ProjectSliceToXY(lats, lons, centerLatRad)

	// Build set of patterns each stop belongs to (for same-pattern-set filtering).
	// We skip transfers between stops that share exactly the same set of patterns,
	// as they're already connected by routes.

	maxDist := float64(config.MaxTransferWalkM)
	walkSpeed := config.WalkSpeedMPerMin

	for i := 0; i < nStops; i++ {
		for j := 0; j < nStops; j++ {
			if i == j {
				continue
			}
			dist := geo.Distance2D(xs[i], ys[i], xs[j], ys[j])
			if dist <= maxDist {
				walkMin := dist / walkSpeed
				transfers[i] = append(transfers[i], Transfer{
					TargetIdx: j,
					WalkMin:   walkMin,
				})
			}
		}
	}

	return transfers
}

// Flatten converts a Timetable into a FlatTimetable with contiguous arrays.
func Flatten(tt *Timetable) *FlatTimetable {
	ft := &FlatTimetable{
		NStops:    tt.NStops,
		NPatterns: tt.NPatterns,
		StopIDs:   tt.StopIDs,
	}

	// Flatten PatternStops -> PSData + PSOffsets
	ft.PSOffsets = make([]int32, tt.NPatterns+1)
	for pi, ps := range tt.PatternStops {
		ft.PSOffsets[pi+1] = ft.PSOffsets[pi] + int32(len(ps))
	}
	ft.PSData = make([]int32, ft.PSOffsets[tt.NPatterns])
	for pi, ps := range tt.PatternStops {
		off := ft.PSOffsets[pi]
		for j, idx := range ps {
			ft.PSData[off+int32(j)] = int32(idx)
		}
	}

	// Flatten PatternTrips -> TTData + TTOffsets + TTNTrips + TTNStops
	ft.TTOffsets = make([]int32, tt.NPatterns+1)
	ft.TTNTrips = make([]int32, tt.NPatterns)
	ft.TTNStops = make([]int32, tt.NPatterns)
	for pi, ptd := range tt.PatternTrips {
		ft.TTNTrips[pi] = int32(ptd.NTrips)
		ft.TTNStops[pi] = int32(ptd.NStops)
		ft.TTOffsets[pi+1] = ft.TTOffsets[pi] + int32(len(ptd.Data))
	}
	ft.TTData = make([]int32, ft.TTOffsets[tt.NPatterns])
	for pi, ptd := range tt.PatternTrips {
		off := ft.TTOffsets[pi]
		for j, v := range ptd.Data {
			ft.TTData[off+int32(j)] = int32(v)
		}
	}

	// Flatten StopToPatterns -> SPData + SPOffsets
	ft.SPOffsets = make([]int32, tt.NStops+1)
	for si, entries := range tt.StopToPatterns {
		ft.SPOffsets[si+1] = ft.SPOffsets[si] + int32(len(entries))
	}
	ft.SPData = make([][2]int32, ft.SPOffsets[tt.NStops])
	for si, entries := range tt.StopToPatterns {
		off := ft.SPOffsets[si]
		for j, e := range entries {
			ft.SPData[off+int32(j)] = [2]int32{int32(e.PatternIdx), int32(e.Position)}
		}
	}

	// Flatten Transfers -> TRData + TROffsets
	ft.TROffsets = make([]int32, tt.NStops+1)
	for si, trs := range tt.Transfers {
		ft.TROffsets[si+1] = ft.TROffsets[si] + int32(len(trs))
	}
	ft.TRData = make([][2]int32, ft.TROffsets[tt.NStops])
	for si, trs := range tt.Transfers {
		off := ft.TROffsets[si]
		for j, tr := range trs {
			ft.TRData[off+int32(j)] = [2]int32{int32(tr.TargetIdx), int32(tr.WalkMin * 100)}
		}
	}

	return ft
}
